warehouse/schema.go

```diff
@@ -1,53 +1,93 @@
 package warehouse
 
-import "fmt"
+import (
```

Comment:
Warning: long-ish comment!

I think we can simplify this a bit and make it more Go like. In the current version, responsibility is spread a bit too much - there's a lot of type mapping here, but then in a specific provider (like `redshift.go`) we're still converting things, looping and mapping to strings, etc. That'll end up being boilerplate that other providers will have to implement as well, so it's nicer to do it once in a single place.

Rough code below is how it looks. Things to point out: the central idea here is that a provider just specifies a Go-to-DB type mapping, eg:

```
var (
	RedshiftTypeMap = FieldTypeMapper{
		"int64":     "BIGINT",
		"string":    "VARCHAR(max)",
		"time.Time": "TIMESTAMP",
	}
)
```
which is passed in to `ExportTableSchema()` to produce a provider-specific schema. (Another advantage of this approach is easy re-use of common type mappers.) 

The Stringers for Field and Schema do the heavy lifting for `create table` type statements - that won't be suitable for every provider, but it's a common case and easy to do here. If something more custom is needed (in BQ it's definitely the case), we can do that in the provider.   

I stripped down the struct types a bit, and removed getters...in Go, it's generally preferable to give direct access to struct fields if it's safe. Also, I put `CustomVars` into `exportSchema` - no need to separate them when formulated this way (it was vestigial from my first version).

There are a couple of ripple changes from this into `main.go`, but it's very minor - we can look at that together.


```
package warehouse

import (
	"fmt"
	"log"
	"reflect"
	"strings"
	"time"
)

type exportSchema struct {
	EventStart             time.Time
	EventType              string
	EventTargetText        string
	EventTargetSelectorTok string
	EventModFrustrated     int64
	EventModDead           int64
	EventModError          int64
	EventModSuspicious     int64
	IndvId                 int64
	PageUrl                string
	PageDuration           int64
	PageActiveDuration     int64
	PageRefererUrl         string
	PageLatLong            string
	PageAgent              string
	PageIp                 string
	PageBrowser            string
	PageDevice             string
	PageOperatingSystem    string
	PageNumInfos           int64
	PageNumWarnings        int64
	PageNumErrors          int64
	SessionId              int64
	PageId                 int64
	UserAppKey             string
	UserEmail              string
	UserDisplayName        string
	UserId                 int64
	CustomVars             string
}

type syncTable struct {
	ID            int64
	Processed     time.Time
	BundleEndTime time.Time
}

type Field struct {
	Name   string
	DBType string
	IsTime bool
}

func (f Field) String() string {
	return fmt.Sprintf("%s %s", f.Name, f.DBType)
}

type Schema []Field

func (s Schema) String() string {
	ss := make([]string, len(s))
	for i, f := range s {
		ss[i] = f.String()
	}
	return strings.Join(ss, ",")
}

type FieldTypeMapper map[string]string

func ExportTableSchema(ftm FieldTypeMapper) Schema {
	return structToSchema(exportSchema{}, ftm)
}

func SyncTableSchema(ftm FieldTypeMapper) Schema {
	return structToSchema(syncTable{}, ftm)
}

func structToSchema(i interface{}, ftm FieldTypeMapper) Schema {
	t := reflect.TypeOf(i)
	result := make(Schema, t.NumField())
	for i := 0; i < t.NumField(); i++ {
		result[i] = Field{
			Name:   t.Field(i).Name,
			DBType: convertType(ftm, t.Field(i).Type),
			IsTime: t.Field(i).Type == reflect.TypeOf(time.Time{}),
		}
	}
	return result
}

func convertType(ftm FieldTypeMapper, t reflect.Type) string {
	dbtype, ok := ftm[t.String()]
	if !ok {
		log.Fatal("Type %s is not present in FieldTypeMapper", t)
	}
	return dbtype
}
```